resourcemanager/taskgroup/task_group.go

```diff
@@ -0,0 +1,223 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package taskgroup
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/tidb/util/grunning"
+	"github.com/pingcap/tidb/util/queue"
+	atomicutil "go.uber.org/atomic"
+)
+
+type key int
+
+// Key is the key of TaskContext
+const Key key = 0
+
+// TaskGroup is a group of tasks
+type TaskGroup struct {
+	cpuTime   atomicutil.Duration
+	startTime time.Time
+
+	// All tasks in this task group is yield and maintained in a link list
+	mu struct {
+		sync.RWMutex
+		tasks *TaskContext
+	}
+}
+
+// TaskContext is the context of a task
+type TaskContext struct {
+	tg *TaskGroup
+	wg sync.WaitGroup
+
+	lastCheck time.Duration
+	next      *TaskContext
+}
+
+// NewTaskGroup is to create a new task group
+func NewTaskGroup() *TaskGroup {
+	return &TaskGroup{
+		startTime: time.Now(),
+	}
+}
+
+func fromContext(ctx context.Context) *TaskContext {
+	val := ctx.Value(Key)
+	if val == nil {
+		return nil
+	}
+	ret, ok := val.(*TaskContext)
+	if !ok {
+		return nil
+	}
+	return ret
+}
+
+// NewContext is to create a new context with task group
+func NewContext(ctx context.Context, tg *TaskGroup) context.Context {
+	return context.WithValue(ctx, Key, &TaskContext{
+		tg: tg,
+	})
+}
+
+// ContextWithSchedInfo is to create a new context with task group
+func ContextWithSchedInfo(ctx context.Context) context.Context {
+	tc := fromContext(ctx)
+	if tc == nil {
+		panic("no TaskContext found in current context!")
+	}
+	return NewContext(ctx, tc.tg)
+}
+
+// CheckPoint is to check point
+func CheckPoint(ctx context.Context) {
+	tc := fromContext(ctx)
+	if tc == nil {
+		return
+	}
+
+	tg := tc.tg
+	tg.mu.RLock()
+	scheduling := tg.mu.tasks != nil
+	tg.mu.RUnlock()
+
+	if !scheduling {
+		runningTime := grunning.Time()
+		elapse := runningTime - tc.lastCheck
+		tc.lastCheck = runningTime
+		cpuTime := tg.cpuTime.Add(elapse)
+		if cpuTime < 20*time.Millisecond {
+			return
+		}
+	}
+
+	tc.wg.Add(1)
+	s.ch <- tc
+	tc.wg.Wait()
+}
+
+type sched struct {
+	ch     chan *TaskContext
+	pq     *queue.PriorityQueue[*TaskGroup]
+	isStop atomic.Bool
+}
+
+func (s *sched) stop() {
+	s.isStop.Store(true)
+}
+
+var s = sched{
+	ch: make(chan *TaskContext, 4096),
+	pq: queue.NewPriorityQueue[*TaskGroup](200, func(a *TaskGroup, b *TaskGroup) bool {
+		return a.startTime.Before(b.startTime)
+	}),
+}
+
+// StopScheduler is to stop the scheduler
+func StopScheduler() {
+	s.stop()
+}
+
+// Scheduler is to schedule the tasks
+func Scheduler() {
+	tg := NewTaskGroup()
+	ctx = NewContext(context.Background(), tg)
+	lastTime := time.Now()
+	const rate = 10
+	capacity := 200 * time.Millisecond
```

Comment:
And capacity is the capacity of the token bucket, which allows burst to some extent.
It's also a value need tuning.

Reply:
I agree with you.